Implement eLearningLab

LiXiaoooowei opened the pull request:
**Key motivations for eLearningLab:**

**From User perspective:**
eLearningLab provides user with an easier way to create/delete/modify caption/callout/voice. Instead of typing text in the speaker note with many tags, user may simply create a `selfExplanationItem` and choose functionalities associated with it.

**From PPTLabs implementation perspective:**
CaptionsLab and NarrationsLab automate the generation of captions and narrations. However, choices are limited when user combine the power of both labs. For example, as a user, what if I would like to generate only narrations for some notes, or to generate only captions for some other notes in the speaker note? What if I would like callout to contain different text from caption? The speaker note syntax would be made more complex to handle the different usage scenarios.

**Features of eLearningLab:**

- Allow callouts to be generated from text which can be different from caption text.
- Automatically infers user preference for newly generated shapes for callouts and captions, i.e. callout shape, callout background, text font, etc. Inference is based on the previous generated shape.
- Support changing of order for self-explanation item by clicking the up and down arrow button.
- Support deletion of self-explanation item by clicking the trash button. If no functionality (i.e. caption, callout, voice) is selected, the self-explanation item will be automatically removed when eLearningLab task pane is reopened on the same slide.
- Support preview of different voices for caption text. User can choose the best voice for text after previewing.
- Sync self-explanation items to animation pane on `sync button` clicked based on the click number sequence as indicated on the item. For example, if a self-explanation item has click number 2, then all its functionalities (i.e. caption, voice, callout) will be added to animation pane under click number 2.

**Special note for eLearningLab:**
- We restrict the number of self-explanation item in one click to be at most 1. That means, if there are two self-explanation items, one followed immediately by another, the `trigger with previous` functionality for the second item is disabled. A diagram illustration is shown in the figure below.  We assume the second item is to be triggered by click in that case. However, if the second item is shifted to a place after custom item, then `trigger with previous` functionality will be enabled again.
- Any user modifications on self-explanation items will only be saved if sync button is pressed. If user changes selection to another slide without clicking the sync button, a message will be shown to user if we detected any unsynced modification. User may choose to sync the changes or discard any changes. 

![picture1](https://user-images.githubusercontent.com/17466278/52833777-97aab000-3119-11e9-83a2-06861c50e149.png)


**Special note during testing:**
We will remind user to sync any unsaved changes if user selection is changed to another slide with unsynced changes on the previous slide. However, if user only modified the text content in text box, changes in text cannot be detected. This can be a bug with underlying implementation, but I haven’t figure it out yet.

LiXiaoooowei commented:
Below is a more detailed illustration of eLearningLab task pane functionalities.

![userguide](https://user-images.githubusercontent.com/17466278/52833206-ba879500-3116-11e9-8c9d-c050190ff747.png)

1.  The up/down/delete button to adjust the self-explanation item's position or delete the item entirely.
2.  Indicate callout/caption/voice functionality. For example, if a callout checkbox is checked, then a callout item will be added to the slide, together with its corresponding animation at click No 0.
3.  Check this place if you would like callout content to be different from caption/voice content. A separate textbox will appear to enter shorter version of callout content. Note that if no text is entered, the program will assume callout text to be the same as caption text.
4. Voice preview button. Note that this button is only enabled if voice functionality next to it is checked.
5. Place to create a new self-explanation item. The newly created item is automatically appended to the back of the list.
6. Sync button, when sync button is pressed, there will be two main operations. Firstly, any custom animations from animation pane will be synced over based on a best effort approach. Secondly, any self-explanation items will be synced over to the animation pane.

LiXiaoooowei commented:
@leeyh20 @blewjy @ChesterSng @YuPeiHenry Ready for review. The code may subject to changes in the future, so let's focus on the functionalities for now. Would any one have time to try out this eLearningLab?

ChesterSng commented:
Hi @LiXiaoooowei great job on Elearning Lab! I have found some bugs and also have some suggestions!

**Bugs**
1. From @blewjy: 
> **Error when closing the window**. This one is a little tricky. But only occurs with ELL, I've tried with our other labs that have panes, and they do not have this problem.
>     _**Steps to reproduce:**_
>     a. Make some changes to the presentation. For e.g. type something into the title box, but don't save your changes.
>     b. Open ELL.
>     c. Close the whole window i.e. Click 'X' on PowerPoint.
>     d. When prompted to save your changes, click 'Cancel'. You should be brought back to PowerPoint.
>     e. Open ELL. You will get an exception here.

This can also happen if you open two PowerPoint windows and then try to open Elearning Lab on both of them.

**Suggestions**
1. > * We restrict the number of self-explanation item in one click to be at most 1. That means, if there are two self-explanation items, one followed immediately by another, the `trigger with previous` functionality for the second item is disabled. A diagram illustration is shown in the figure below.  We assume the second item is to be triggered by click in that case. However, if the second item is shifted to a place after custom item, then `trigger with previous` functionality will be enabled again.

This part is a bit confusing because I still see the `With Previous` option selected just that I am unable to change it. I will thus assume that the `self-explanation item` is in the state `With Previous` and is unchangeable.  Same with in Click 0, it shows `On Click` and greyed out. Maybe it can be a blank and greyed out?

2. Maybe `Use shorter callouts` can be renamed to `Use Alternative Callouts`? It might not necessarily be shorter!

3. Currently the implementation is that Click 0 will be shown when the slide is opened in slide show. However, if the user wants the first callout/voice/caption to appear after one click, the user can't do it easily with Elearning Lab. The above behaviour can be achieved in this way though:
a. Create Self Explanation Item (It is click 0)
b. Add custom animation
c.  Sync
d. Self Explanation Item from a. is now click 1.

4. The settings button can be shifted to the pane so that Elearning Lab opens with one click instead of two.

5. Up and Down buttons can be disabled for first and last `self-explanation item`

LiXiaoooowei commented:
Thanks for testing. I have made some modifications based on your feedback and also corrected some of the bugs reported.


@blewjy Bugs fixed. I also change the `onslideselectionchange` behavior of ELL so that it automatically refreshes upon slide selection change.

> 2. **My items are lost when I change slides.** This only occurs for items that do not have any generated caption, callout, or audio, but just a new item that is sitting in the pane. After changing slides and coming back, the item is lost. 

If there are no functionalities associated with an item, then the item is considered as deleted.  This is the actually the intended behavior for now.


@ChesterSng 

Bugs fixed on point 1. 

UI updated on point 2. 

Points 3 and 4 may be considered as enhancements that I will discuss further with prof. 

For point 5 on up and down buttons, since the UI is not fixed yet, I will temporarily leave them at their current state for now.


@YuPeiHenry 

I am not able to reproduce point 2 and 5. Would you check again to see if the same exception still appear?

Regarding point 4, I will discuss further with Prof. The current implementation assumes callout text to be the same as caption text if callout text is empty. That's why you still see "add" in the textbox.  

Regarding point 6, the undo behavior sometimes work as expected in reverse order, but sometimes it will simply undo the all the changes performed by ELL. For now, I will leave "undo" to the default PPT's implementation. I might enhance the undo behavior in the future.

YuPeiHenry commented:
@LiXiaoooowei Error 2 can still be replicated on PPT2010, does not seem to happen on PPT2016 though. Might want to ignore that for now since it only affects PPT2010. Error 5 have been fixed (I could replicate it on PPT2016 previously, just confirmed).

blewjy commented:
Hey @LiXiaoooowei, there seems to be still a little problem with the Azure voice and generating a voice narration for my items. Even though I'm not logged in to the Azure voice and I'm using the default/In-Computer voice, it still prompts me with a pop-up saying that the azure service cannot be authenticated.

***Steps to reproduce:***
a. Make sure you are not logged into the Azure speech service/Cloud voice.
b. Create a new item, tick 'Voice' and click 'Sync'
c. There will be a popup saying Azure Account Authentication Failed.

Another small issue is that user is allowed to type in multi-line notes when creating a new item in the textbox near the bottom of the pane, but after the item is created, user is not able to create multiple lines using the enter key. I think the behaviour should be consistent between these two textboxes -- either allow or disallow multi-line notes for both.

Lastly, will there be any tests written for this new lab?

LiXiaoooowei commented:
@YuPeiHenry That is the current implementation because unlike callout text, audio is generated based on caption text. Therefore, if caption text is deleted, audio will not be generated. 

@blewjy I have modified the code so that `Azure account invalid` message will only be prompted to user if azure voice is chosen in at least one of the self explanatory items. Textbox UI also updated to ensure consistancy.

ChesterSng commented:
Hi @LiXiaoooowei did some further testing:

**Processing speech box appears for a brief moment even if there was no voice selected.**

_Steps to reproduce:_
1. Add self-explanation item with callout/captions with no voice. 
2. Sync 

![image](https://user-images.githubusercontent.com/39242483/53071393-0ecab480-351d-11e9-92a3-f567b35083c8.png)

**Empty input allowed for shorter callouts**
_Note that in step 3, if nothing is typed, after syncing, the shorter callout will be unchecked automatically._

_Steps to reproduce:_
1. Add self-explanation item with callout. 
2. Type something into textbox
2. Tick `use shorter callouts`
3. Type something into shorter callout textbox and then delete whatever is typed 
4. Sync (No warning)
5. Callout becomes empty as well 
6. Close eLearning Lab (Unsynced item warning will appear, selecting yes or no doesn't matter)
7. Callout remains empty
8. Open eLearning Lab (Shorter callout will be unchecked)
9. Sync (Callout becomes populated again)

**Similar to above, but affects captions selection ability**

_Steps to reproduce:_
1. Add self-explanation item. 
2. Tick `use shorter callouts`
3. Type something and then delete whatever is typed
4. Tick Captions
5. Sync
6. Captions is automatically unchecked

**Captions overflows slide when there are 3 lines and above**

_Steps to reproduce:_
1. Add self-explanation item with captions. 
2. Add at least 3 lines as text
3. Overflows slide
![image](https://user-images.githubusercontent.com/39242483/53072638-b85f7500-3520-11e9-8a9c-54a04f01409f.png)

LiXiaoooowei commented:
Hi @ChesterSng Point 1, 2 and 3 are the current implementation. During syncing, I will check if there are empty callout content / caption content, and untick isCaption/isCallout/isVoice if its corresponding text is empty. Because there is no point of generating empty content. As for point 4, if there are multiple lines in the caption, then caption box should display those multiple lines as well right? Would you have a better suggestion?

blewjy commented:
@LiXiaoooowei I think what @ChesterSng meant was that the caption box generated overflows the slide and cannot be seen completely when in slideshow mode. His screenshot shows that "item 3" and "item 4" are not within the slide boundaries.

That being said, I'm unable to reproduce this error. Multi-line caption box when generated are within the slide boundaries for me. Maybe @ChesterSng you wanna try again to see if it's working for you?

ChesterSng commented:
Hi @LiXiaoooowei I think you misunderstood points 2 and 3, it is not about empty callout/caption content. It is the `Use Shorter Callout` textbox.

This is the correct and implemented behaviour: 
1. Tick `Use Shorter Callout`
2. Enter nothing into the `Shorter Callout textbox`
3. Sync
4. **`Use Shorter Callout` becomes unchecked**

However, if we do these steps instead: 
1. Tick `Use Shorter Callout`
2. Type something into `Shorter Callout Textbox`
3. Delete everything in `Shorter Callout Textbox`
4. Sync
5. **Does not uncheck `Use Shorter Callout`**

This will result in bugs 2 and 3. 

Also @blewjy is right in interpreting bug 4. It overflows the slide, tested on powerpoint 2016.

LiXiaoooowei commented:
@ChesterSng Sorry for misreading your points. I fixed point 2, 3 and 4. Code will be pushed later after some more refinements. Moreover,  I also realized that the tick/untick of checkboxes and how callout text is updated may be not so intuitive at times.  For example,  if callout text is empty, shall I populate caption text instead? But what if caption text is also empty? I will make a list specifying the expected behavior of all those gray areas, i.e. how checkboxes are updated with callout/caption text as well as how callout text is updated with caption text.

LiXiaoooowei commented:
@ChesterSng Problem fixed. 

The following behaviors apply to CalloutText and CaptionText (tentative):
- By default (when `Use Shorter Callout` not checked), callout text will follow caption text. Any changes in caption text will be applied to callout text as well.
- If `Use Shorter Callout` box is checked and `CalloutTextBox` is non-empty, contents in `CalloutTextBox` will be updated to slide.
- If `Use Shorter Callout` box is checked but `CalloutTextBox` is empty, `IsCallout` will be unchecked and callout shape on slide will be deleted.
- If `CaptionTextBox` is empty when `Use Shorter Callout` is selected, then `IsCaption` will be unchecked and  caption on slide will be deleted.
- If `CaptionTextBox` is empty when `Use Shorter Callout` is not selected, then both `IsCaption` and `IsCallout` will be unchecked, and any callouts or captions will be deleted.

@YuPeiHenry @blewjy @ChesterSng  I changed the storage method for elearning lab text so that all the text will now be stored to slide in a hidden shape, therefore, elearninglab is able to recover selfExplanationItem even if none of `isCaption`, `isCallout` and `isVoice` is applied to the item. 

The recovered self explanation item will not have any click number associated with it. Note that eLearningLab is able to restore the relative position for such self explanation item among other self explanation items. Its exact position w.r.t custom animations is not guaranteed. 

Moreover, since the only source of text for callouts/captions/voice should come from eLearningLab (which is stored in hidden shape on slide), any changes in callout/caption text from the slide directly will not be synced to eLearningLab. If users want to change callout/caption content, they should do that from eLearningLab rather than perform the operation on slides directly.

ChesterSng commented:
Hi @LiXiaoooowei tested the new changes on PowerPoint 2016. 

> * By default (when `Use Shorter Callout` not checked), callout text will follow caption text. Any changes in caption text will be applied to callout text as well.
> * If `Use Shorter Callout` box is checked and `CalloutTextBox` is non-empty, contents in `CalloutTextBox` will be updated to slide.
> * If `Use Shorter Callout` box is checked but `CalloutTextBox` is empty, `IsCallout` will be unchecked and callout shape on slide will be deleted.
> * If `CaptionTextBox` is empty when `Use Shorter Callout` is selected, then `IsCaption` will be unchecked and  caption on slide will be deleted.
> * If `CaptionTextBox` is empty when `Use Shorter Callout` is not selected, then both `IsCaption` and `IsCallout` will be unchecked, and any callouts or captions will be deleted.

The behaviours that you listed are working correctly!
___
**1. Syncing on an empty presentation will cause an unhandled exception to be thrown**
i. Select a slide
ii. Opening eLearning Lab 
iii. Sync 
iv. Select slide from left panel 
v. Error appears on Visual Studio
_If I run Powerpoint from my desktop instead of visual studio, then this error will not occur. i.e. Does not affect users_ 
![image](https://user-images.githubusercontent.com/39242483/53334186-02869300-3933-11e9-9e35-84498ce09f0a.png)
___
**2. Inconsistency in deletion of self-explanation item**
i. Create a self explanation item (check callout, captions and voice) 
ii. Delete the self explanation item (**only callout and captions is deleted**, voice button still visible on slide)
iii. Sync (will delete the voice button)

i. Create a self explanation item (check callout, captions and voice) 
ii. Delete the self explanation item (**only callout and captions is deleted**, voice button still visible on slide)
iii. Go into slide show mode (F5, voice is activated)
iv. Exit slide show mode, self explanation item appears with voice enabled. 
_Step iii. here can be replaced with tabbing to another slide and back as well_
___
**3. No warning to save items that haven't been synced**
_Not sure about this, but if I recall correctly, if I make an self-explanation item and check any of the boxes, on changing of slides, I will be prompted to sync my changes_
i. Create a self-explanation item (check any of callout/captions/voice, don't sync)
ii. Change slides (No warning)
iii. Return to original slide (self-explanation item will be gone)
___
**4. I previously posted this issue, but not sure if I phrased it properly.**
__Intended behaviour is that the first item in eLearningLab will be Click 0__
_Following these steps, I can get the first item to be Click 1_ 
i. Add a self explanation item (check any of callout/captions/voice) 
ii. Add a self explanation item (don't check anything)
iii. Reorder the blank self explanation item (in step 2), and then sync (Notice the self explanation item in step 1 is already Click 1) 
iv. Delete the blank self explanation item and sync (Self explanation item in step 1 is still Click 1)
![click 1](https://user-images.githubusercontent.com/39242483/53332947-7b83eb80-392f-11e9-8578-deefd417eee1.gif)

LiXiaoooowei commented:
Hi @ChesterSng Thanks for testing! I am able to reproduce issue 2 and 4. However, I couldn't really reproduce issue 1 and 3. Maybe you can try again?

ChesterSng commented:
Hi @LiXiaoooowei no problem! I am able to reproduce points 1 and 3 reliably on another machine (PowerPoint 2013) as well though. 

Here is point 1: 

![syncing nothing](https://user-images.githubusercontent.com/39242483/53350309-cadf1180-3959-11e9-8507-a9a5b9a77b7b.gif)

Here is point 3:

![no warning on not syncing](https://user-images.githubusercontent.com/39242483/53350454-0f6aad00-395a-11e9-938c-86b7c85e433c.gif)

LiXiaoooowei commented:
@ChesterSng I can reproduce point 1 now. For point 3, the issue happens on my computer only if you create and edit a new item that does not previously exists in elearninglab. If I edit an item that already exists in elearninglab (synced before), then the sync message will show up properly.  Is it the same behavior on your machine as well?

ChesterSng commented:
@LiXiaoooowei Yes for an existing item, the sync message will show up properly

LiXiaoooowei commented:
@ChesterSng Fixes 1 and 3. 

For point 2, I feel that deleting callout/caption/audio shapes immediately when an item is deleted involves iterating over all shapes on the slide. If user deletes several items, then the code behind will iterate over all shapes multiple times. Since our policy is that the slide will only be synced with elearninglab after `sync` button is pressed, there is minimal gain to delete callout/audio/caption shapes immediately. To enforce consistency for item deletion, I hold on the deletion of callout/caption until `sync` button is pressed. 

For point 4, I haven't come up with a neat and clean way to update click number for self explanation item if the previous self explanation item does not have any functionalities (isCaption/isVoice/isCallout). I will probably address this issue again in the future.

blewjy commented:
Hi @LiXiaoooowei, I tested this on PPT 2013.

**1. Previously synced settings not saved if I were to start a slideshow**
a. Create a new item, tick callout/caption/voice, but don't click `Sync`.
b. Start slideshow immediately (press F5)
c. There will be nothing on the slideshow. This is expected, because items are not synced yet.
d. When slideshow ends, popup will appear, prompting to sync items. Click `Yes`
e. the callout/captions/voice will be generated on the slide, but in the ELL pane, the settings are not checked.
f. reopening the ELL pane or going to slideshow mode and out again will fix this, but if user were to click on `Sync` when the settings are unchecked, it will cause the user to lose the generated captions, text and audio.
***Expected behaviour:*** Relevant settings should be checked at step (e).

**2. E-learning Lab settings (and also Narration Lab settings): many MessageBoxes when clicking in and out of the login page**
a. Open ELL or Narration Lab settings and login to Azure speech service.
b. On your first selection of `Cloud Voice`, there should be a popup saying that "only english is supported". Dismiss this popup.
c. Click `Click to change user account`, then click `Cancel` on the login page popup.
d. select `In-Computer Voice`, then select `Cloud Voice` again. You will encounter multiple popups (at this point, it is 3 repeated popups for me).
e. if you repeat this action, you will continue to get more and more popups.
![narration-lab-settings-popup](https://user-images.githubusercontent.com/39292809/53475859-ff62e280-3aab-11e9-8591-1a64f0b633c4.gif)

**3. Is there no way for user to show the first item on the first click instead of it being click 0?** _(Just for your consideration, not really a bug)_
Related to:
> **4. I previously posted this issue, but not sure if I phrased it properly.**
**Intended behaviour is that the first item in eLearningLab will be Click 0**
_Following these steps, I can get the first item to be Click 1_
i. Add a self explanation item (check any of callout/captions/voice)
ii. Add a self explanation item (don't check anything)
iii. Reorder the blank self explanation item (in step 2), and then sync (Notice the self explanation item in step 1 is already Click 1)
iv. Delete the blank self explanation item and sync (Self explanation item in step 1 is still Click 1)

It might be useful to allow the user to set the first item to only show on Click 1, instead of restricting it to Click 0 only for the first item. I don't know how difficult it would be to implement this. However, I think it's something to keep in mind.


**4. Deletion of item**
Related to:
> **2. Inconsistency in deletion of self-explanation item**
i. Create a self explanation item (check callout, captions and voice)
ii. Delete the self explanation item (**only callout and captions is deleted**, voice button still visible on slide)
iii. Sync (will delete the voice button)<br>
i. Create a self explanation item (check callout, captions and voice)
ii. Delete the self explanation item (**only callout and captions is deleted**, voice button still visible on slide)
iii. Go into slide show mode (F5, voice is activated)
iv. Exit slide show mode, self explanation item appears with voice enabled.
_Step iii. here can be replaced with tabbing to another slide and back as well_

a. Create a new item, tick caption, callout and voice, then click sync.
b. Delete the item with the trash bin button
c. From Chester's bug report (as quoted above) it seems that the caption and callout will be immediately deleted. But this is not the case for me (PPT 2013). The caption, callout and voice still remains on the slide, until i click `Sync`
e. when `Sync` button is clicked, the callout and voice gets removed, but the caption remains.
![ell-item-deletion](https://user-images.githubusercontent.com/39292809/53476032-6b454b00-3aac-11e9-9f61-f9672c221cef.gif)

**Some minor UI-related issues to take note. Probably not urgent to fix, but should ultimately be fixed.**
1. Ribbon button is "e-Learning", but the pane title and button title is "eLearningLab". Probably want to keep it consistent with the title for other labs, and name it "e-Learning Lab".
2. Voice checkbox and the audio preview button is not aligned with the rest of the checkboxes.

LiXiaoooowei commented:
@blewjy Hi Bryan, thanks for testing! I have fixed the points you mentioned above, except for point 2 and UI tweaks 1.  For point 2, currently `only English language supported` message is prompted every time the `azure voice radio button` is checked. I will address this issue and UI tweaks 1 in a separate issue in the future.

ChesterSng commented:
Hi @LiXiaoooowei, tested your new changes! 

**First item can now be "click 2"**
_Steps to reproduce:_
i. Make a self-explanation item (tick callout and enter some text)
ii. Change it to `On Click` and Sync
iii. Make a self explanation item (don't tick anything but enter some text)
iv. Shift it to the top (self-explanation item 1)
v. Change it to `On Click`
vi. Sync
vii. Enter Slide Show mode
_Some weird behaviours will follow on click 2, in the GIF, two self-explanation items are used to show the weird behaviours._
![click 2](https://user-images.githubusercontent.com/39242483/53504362-c6972d80-3aec-11e9-8e0e-71d6fd76bed7.gif)


Have a suggestion for you to consider:
 
**Confusing option in disabled drop down** 
As seen in the screenshot below, self-explanation item 2 and 3 have `With Previous` greyed out. 
So in this case, are self-explanation items click 2 and 3? Or do they appear `With Previous` (meaning they appear together with self explanation item 1)? `On Click` greyed out might be more accurate and less confusing.
_I think it users will think that the greyed out option is not that is __disabled__ but actually __fixed__ to that greyed out option_
![image](https://user-images.githubusercontent.com/39242483/53503024-3a840680-3aea-11e9-9ffa-767257900eaf.png)

leeyh20 commented:
I agree that the greyed out option looks like it is fixed to that option (the option cannot be changed) rather than disabled. If the option is disabled, could the option should not be included in the drop down?

leeyh20 commented:
> For point 2, I feel that deleting callout/caption/audio shapes immediately when an item is deleted involves iterating over all shapes on the slide. If user deletes several items, then the code behind will iterate over all shapes multiple times. Since our policy is that the slide will only be synced with elearninglab after `sync` button is pressed, there is minimal gain to delete callout/audio/caption shapes immediately. To enforce consistency for item deletion, I hold on the deletion of callout/caption until `sync` button is pressed.

Is there a clear way to indicate to the user that they have some changes and that they should actually press the sync button? Should the sync button be only activated if there are some changes made (able to be synced) and at other times it is disabled?

LiXiaoooowei commented:
@ChesterSng Bug fixed.

@ChesterSng @leeyh20 For the greyed out trigger type button, the current behavior is that greye- out self explanations will always be triggered on click. The displayed content on greyed-out button (onClick or With previous) is based on its previous active display. Since we do not allow two self explanations to appear under the same click number, therefore, if a self explanation item is pushed up/down such that it follows another self explanation item, then its trigger type button will be greyed-out. However, if the position of the self explanation item moves such that it follows a custom item instead, then its previous active trigger type will be effective. For other ways of implementation, let's address in a separate issue.

LiXiaoooowei commented:
@leeyh20 Yes a message which prompts user to sync will be displayed to user when user changes slide selection. This message will only be displayed if user made any modifications to elearninglab items. `Sync` button will always be available.

damithc commented:
In general, we prefer to disable non-applicable UI item rather than hide them. The rationale is that it improves learnability:
1. the user will see the disabled item and will realize that the option exists but not available for some reason
2. the UI will remain the same all the time. Hiding elements can make the UI morph into different forms at different times.
But of course there are downsides and we can make exceptions to that policy when necessary.

LiXiaoooowei commented on PowerPointLabs/PowerPointLabs/Views/ProcessingStatusForm.cs:
```diff
@@ -1,39 +0,0 @@
-﻿using System.Windows.Forms;
-
-namespace PowerPointLabs.Views
-{
```
The original ProcessingStatusForm is written in WinForm. Since eLearningLab also needs to display ProgressingStatusBar to show progress, so I rewrote `ProcessingStatusForm` in WPF and deleted the original.

leeyh20 replied:
Oh cool, thus it affects this too, I see.

LiXiaoooowei commented on PowerPointLabs/PowerPointLabs/ELearningLab/Views/AudioSettingsViews/SpeechPlayingDialogBox.xaml.cs:
```diff
@@ -40,6 +43,10 @@ private void SynthesizerOnSpeakCompleted(object sender, SpeakCompletedEventArgs
         {
+           // if (player != null)
+        //    {
+         //       player.Stop();
```
Removed!

LiXiaoooowei commented on PowerPointLabs/PowerPointLabs/ELearningLab/AudioGenerator/AzureVoiceGenerator/AzureVoiceSpeechEngine/SynthesizeAzureVoice.cs:
```diff
@@ -100,6 +115,72 @@ public Task Speak(CancellationToken cancellationToken, InputOptions inputOptions
+                return null;
+            }
+
+            var saveTask = httpTask.ContinueWith(
```
Removed duplicate methods (line 87-94)

LiXiaoooowei commented on PowerPointLabs/PowerPointLabs/ELearningLab/Service/AudioRuntimeService/ComputerVoiceRuntimeService.cs:
```diff
@@ -66,57 +95,46 @@ public static string[] EmbedCurrentSlideNotes()
-
             List<PowerPointSlide> slides = PowerPointCurrentPresentationInfo.SelectedSlides.ToList();
-            if (NotesToAudio.IsAzureVoiceSelected && UserAccount.GetInstance().IsEmpty())
+         //   AudioService.SetTempName();
```
Removed!

LiXiaoooowei commented on PowerPointLabs/PowerPointLabs/ELearningLab/Utility/ShapeUtility.cs:
```diff
@@ -0,0 +1,154 @@
+{
+    public class ShapeUtility
+    {
+#pragma warning disable 0618
```
`SlideWith` and `SlideHeight` information are retrieved from `PowerPointCurrentPresentationInfo` when creating shapes. Since ShapeUtility does not inherit from ActionFramework and it is meant to be a utility class so that other services class can use this class and create shapes to slide, I find it a bit hard to use ActionFramework in this case.
